crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -40,24 +46,160 @@ pub use bevy_ecs_macros::States;
 /// }
 ///
 /// ```
-pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {
-    /// The type returned when iterating over all [`variants`](States::variants) of this type.
-    type Iter: Iterator<Item = Self>;
+pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {}
+
+/// Types that can match world-wide states.
```

Comment:
This could be clearer. I'd like more information about the types it's used with, the blanket impl and how it differs from `PartialEq`.

A doc test with a simple and advanced implementation would also be good.